app/org/sagebionetworks/bridge/cache/CacheProvider.java

```diff
@@ -127,24 +114,92 @@ public void setUserSession(final UserSession session) {
+        }
+        return session;
+    }
+    
+    private UserSession getUserSessionNewVersion(String sessionToken) {
+        checkNotNull(sessionToken);
+        try {
+            CacheKey tokenToUserIdKey = CacheKey.tokenToUserId(sessionToken);
+            String userId = jedisOps.get(tokenToUserIdKey.toString());
+            if (userId != null) {
+                CacheKey userIdToSessionKey = CacheKey.userIdToSession(userId);
+                String ser = jedisOps.get(userIdToSessionKey.toString());
+                if (ser != null) {
+                    UserSession session = BridgeObjectMapper.get().readValue(ser, UserSession.class);
+                    // The token --> userId look up is not replaced on session invalidation. 
+                    // Check here and only return if the sessionToken is valid. It is possible 
+                    // to successfully sign in and then have this fail due to concurrent requests.
+                    // The client needs to manage concurrent requests if it doesn't want to 
+                    // invalidate its own session.
+                    if (session.getSessionToken().equals(sessionToken)) {
+                        return session;
+                    }
+                    // Otherwise, delete the key sessionToken key (it's known to be invalid)
+                    jedisOps.del(tokenToUserIdKey.toString());
+                }
+            }
+            return null;
+        } catch (Throwable e) {
+            promptToStartRedisIfLocal(e);
+            throw new BridgeServiceException(e);
+        }
+    }
+    
+    /**
+     *  This is the original method for saving a session which we can leave in place until 
+     *  all sessions have been rotated to storage under the new keys (a day will do).
+     */
+    @SuppressWarnings("deprecation")
+    private UserSession getUserSessionOldVersion(final String sessionToken) {
         checkNotNull(sessionToken);
         try {
             final CacheKey sessionKey = CacheKey.session(sessionToken);
             String ser = jedisOps.get(sessionKey.toString());
             if (ser == null) {
                 return null;
             }
-            return bridgeObjectMapper.readValue(ser, UserSession.class);
+            return BridgeObjectMapper.get().readValue(ser, UserSession.class);
         } catch (Throwable e) {
             promptToStartRedisIfLocal(e);
             throw new BridgeServiceException(e);
         }
     }
-
-    public UserSession getUserSessionByUserId(final String userId) {
+    
+    public UserSession getUserSessionByUserId(String userId) {
+        UserSession session = getUserSessionByUserIdNewVersion(userId);
+        if (session == null) {
+            session = getUserSessionByUserIdOldVersion(userId);
+        }
+        return session;
+    }
+    
+    private UserSession getUserSessionByUserIdNewVersion(final String userId) {
+        checkNotNull(userId);
```

Comment:
Similarly, we should add unit tests for this as well, including the following cases:

* user ID has no session
* user ID has session